### PR TITLE
More csrf and cookies

### PR DESCRIPTION
**Pull request by dogeared**
Added more csrf and cookie support where needed. Shouldn't impact other integrations/languages.

Added a method that will return an account href when an email and password is passed in to delete accounts created by registering.

### CODE REVIEW

File: src/test/groovy/com/stormpath/tck/AbstractIT.groovy [hunk 1]
```diff
@@ -134,6 +146,64 @@ abstract class AbstractIT {
         csrf = ret
     }
 
+    protected void setCSRFAndCookies(requestSpecification, contentType) {
+
+        if (Strings.hasText(csrf) && ContentType.JSON.equals(contentType)) {
+            requestSpecification.header("X-CSRF-TOKEN", csrf)
```
Comment (edjiang) [inline src/test/groovy/com/stormpath/tck/AbstractIT.groovy:152]: @robertjd knows better than me, but should we be also set the `X-XSRF-TOKEN` header here?

File: src/test/groovy/com/stormpath/tck/authentication/BearerIT.groovy [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import com.stormpath.tck.AbstractIT
+import org.testng.annotations.Test
+
+import static com.jayway.restassured.RestAssured.given
+import static com.stormpath.tck.util.FrameworkConstants.MeRoute
+
+class BearerIT extends AbstractIT {
+
+    @Test(groups=["v100", "json", "html"])
+    public void refreshTokenAsAccessBearerTokenFails() throws Exception {
```
Comment (edjiang) [inline src/test/groovy/com/stormpath/tck/authentication/BearerIT.groovy:12]: Isn't this covered here? 

https://github.com/stormpath/stormpath-framework-tck/blob/adc6b64a1bfbee0b71b45772dfc695678df42fb1/src/test/groovy/com/stormpath/tck/authentication/CookieIT.groovy#L149

File: src/test/groovy/com/stormpath/tck/authentication/CookieIT.groovy [hunk 3]
```diff
@@ -91,54 +73,43 @@ class CookieIT extends AbstractIT {
...
     }
 
-    /** Reject unauthorized text/html requests with 302 to login route
-     * 302 to login route should preserve original URL including query string
-     * @see <a href="https://github.com/stormpath/stormpath-framework-tck/issues/36">#36</a>
-     * @see <a href="https://github.com/stormpath/stormpath-framework-tck/issues/121">#121</a>
-     * @throws Exception
-     */
-    @Test(groups=["v100", "html"])
-    public void unauthorizedHtmlRequestIsForwardedToLogin() throws Exception {
```
Comment (edjiang) [inline src/test/groovy/com/stormpath/tck/authentication/CookieIT.groovy:103]: This is a correct test as per our hipchat conversation, and should not be removed.

File: src/test/groovy/com/stormpath/tck/authentication/CookieIT.groovy [hunk 4]
```diff
@@ -91,54 +73,43 @@ class CookieIT extends AbstractIT {
...
+                .post(LoginRoute)
+            .then()
+                .statusCode(200)
+            .extract()
+                .response()
 
         def accessTokenCookie = response.detailedCookies.get("access_token")
         def accessTokenTtl = JwtUtils.extractJwtClaim(accessTokenCookie.value, "exp") as long
-        assertEquals(accessTokenCookie.expiryDate, new Date(accessTokenTtl * 1000L))
+        assertTrue accessTokenTtl * 1000L - accessTokenCookie.expiryDate.time < 2000
```
Comment (edjiang) [inline src/test/groovy/com/stormpath/tck/authentication/CookieIT.groovy]: Why should there be drift here? Both the cookie expires value and the JWT expires value are set in absolute terms, so absolutely should be equal.

File: src/test/groovy/com/stormpath/tck/util/CookieUtils.groovy [hunk 5]
```diff
@@ -22,6 +22,6 @@ class CookieUtils {
 
     public static boolean isCookieDeleted(Cookie cookie) {
         def now = new Date()
-        return cookie.expiryDate < now
+        return cookie == null || cookie.expiryDate < now
```
Comment (edjiang) [inline src/test/groovy/com/stormpath/tck/util/CookieUtils.groovy]: If cookie is null, it is not deleted. Cookies can only be deleted by setting the expiry date to before the current time. 

https://msdn.microsoft.com/en-us/library/ms178195.aspx

File: src/test/groovy/com/stormpath/tck/AbstractIT.groovy (hunk 1 above)
Comment (dogeared) [inline src/test/groovy/com/stormpath/tck/AbstractIT.groovy:152]: @robertjd - should I add that in to this method?

File: src/test/groovy/com/stormpath/tck/authentication/BearerIT.groovy (hunk 2 above)
Comment (dogeared) [inline src/test/groovy/com/stormpath/tck/authentication/BearerIT.groovy:12]: Yes. I removed it from that method. It didn't seem right to have a Bearer test in a test meant to exercise cookie operations.

File: src/test/groovy/com/stormpath/tck/authentication/CookieIT.groovy (hunk 3 above)
Comment (dogeared) [inline src/test/groovy/com/stormpath/tck/authentication/CookieIT.groovy:103]: I don't think @omgitstom ever weighed in on this. It doesn't make sense (to me) in the context of a JSON _only_ endpoint to redirect to an HTML endpoint.

File: src/test/groovy/com/stormpath/tck/authentication/CookieIT.groovy (hunk 4 above)
Comment (dogeared) [inline src/test/groovy/com/stormpath/tck/authentication/CookieIT.groovy]: Not sure about this. The JSDK was consistently off by some amount of time and a number of people weighed in on this and thought that some drift should be accommodated.

File: src/test/groovy/com/stormpath/tck/util/CookieUtils.groovy (hunk 5 above)
Comment (dogeared) [inline src/test/groovy/com/stormpath/tck/util/CookieUtils.groovy]: If there is no cookie present at all, such as on an initial request, the cookie will be null.

**Comment by dogeared** [comment]
@edjiang - I rebased from master, fyi.

File: src/test/groovy/com/stormpath/tck/forgot/ChangePasswordIT.groovy [hunk 6]
```diff
@@ -168,12 +176,18 @@ class ChangePasswordIT extends AbstractIT {
...
-        given()
+        saveCSRFAndCookies(ForgotRoute)
+
+        def req = given()
             .accept(ContentType.HTML)
             .contentType(ContentType.URLENC)
             .queryParam("sptoken", "NOTEVENCLOSETOVALID")
             .param("password", "foobar123!")
-        .when()
+            .param("confirmPassword","foobar123!")
```
Comment (dogeared) [inline src/test/groovy/com/stormpath/tck/forgot/ChangePasswordIT.groovy]: @edjiang @robertjd @nbarbettini @bretterer - will providing the `confirmPassword` field hurt other integrations? I wouldn't think so, but I wanted to check. The backend password reset form and all of the Java Integrations require password confirmation.

File: src/test/groovy/com/stormpath/tck/AbstractIT.groovy (hunk 1 above)
Comment (edjiang) [inline src/test/groovy/com/stormpath/tck/AbstractIT.groovy:152]: @robertjd?

File: src/test/groovy/com/stormpath/tck/AbstractIT.groovy [hunk 7]
```diff
@@ -134,6 +146,64 @@ abstract class AbstractIT {
...
+        if (Strings.hasText(csrfKey) && Strings.hasText(csrf) && !ContentType.JSON.equals(contentType)) {
+            requestSpecification.formParam(csrfKey, csrf);
+        }
+
+        if (cookies != null) {
+            requestSpecification.cookies(cookies)
+        }
+    }
+
+    protected String getAccountHref(String email, String password) {
```
Comment (edjiang) [inline src/test/groovy/com/stormpath/tck/AbstractIT.groovy]: No need to do this, see TestAccount.href:

https://github.com/stormpath/stormpath-framework-tck/blob/master/src/test/groovy/com/stormpath/tck/util/TestAccount.groovy#L23

File: src/test/groovy/com/stormpath/tck/authentication/BearerIT.groovy (hunk 2 above)
Comment (edjiang) [inline src/test/groovy/com/stormpath/tck/authentication/BearerIT.groovy:12]: Cool =]

File: src/test/groovy/com/stormpath/tck/AbstractIT.groovy [hunk 8]
```diff
@@ -134,6 +146,64 @@ abstract class AbstractIT {
...
+            .then()
+                .statusCode(200)
+                .contentType(ContentType.JSON)
+            .extract()
+                .response().getBody()
+
+        return body.jsonPath().get("account.href");
+    }
+
+    protected Tuple2 createTestAccountTokens() {
```
Comment (edjiang) [inline src/test/groovy/com/stormpath/tck/AbstractIT.groovy:164]: I would hesitate to do this in AbstractIT; maybe in TestAccount.groovy?

File: src/test/groovy/com/stormpath/tck/authentication/CookieIT.groovy [hunk 9]
```diff
@@ -120,27 +99,36 @@ class CookieIT extends AbstractIT {
     @Test(groups=["v100", "json"])
     public void cookieExpirationMatchesTokenTtl() throws Exception {
         def account = new TestAccount()
+        saveCSRFAndCookies(FrameworkConstants.RegisterRoute)
```
Comment (edjiang) [inline src/test/groovy/com/stormpath/tck/authentication/CookieIT.groovy]: I would suggest that you put this logic in TestAccount.groovy

File: src/test/groovy/com/stormpath/tck/authentication/CookieIT.groovy (hunk 4 above)
Comment (edjiang) [inline src/test/groovy/com/stormpath/tck/authentication/CookieIT.groovy]: Let's discuss this drift directly. I am still unsure about why this is in the TCK.

File: src/test/groovy/com/stormpath/tck/forgot/ChangePasswordIT.groovy (hunk 6 above)
Comment (edjiang) [inline src/test/groovy/com/stormpath/tck/forgot/ChangePasswordIT.groovy]: This should not be in the tests; password confirmation should be disabled by default:

https://github.com/stormpath/stormpath-framework-spec/blob/20a7dfc52a95ee45969fae59aaa3f3bd161659db/example-config.yaml#L101

File: src/test/groovy/com/stormpath/tck/register/RegisterIT.groovy [hunk 10]
```diff
@@ -509,22 +510,30 @@ class RegisterIT extends AbstractIT {
...
+        setCSRFAndCookies(req, ContentType.HTML)
+
+        req
+            .when()
+                .post(RegisterRoute)
+            .then()
+                .statusCode(302)
+                .header(
+                    "Location",
+                    both(containsString(FrameworkConstants.LoginRoute)).and(containsString("?status=created"))
```
Comment (edjiang) [inline src/test/groovy/com/stormpath/tck/register/RegisterIT.groovy]: What is this testing for that's different?

File: src/test/groovy/com/stormpath/tck/util/CookieUtils.groovy (hunk 5 above)
Comment (edjiang) [inline src/test/groovy/com/stormpath/tck/util/CookieUtils.groovy]: That's true, but this is asserting that the cookie is deleted. If we're expecting a cookie to be actively deleted, it should not be null.

**Comment by edjiang** [comment]
Here's a PR isolating registration so it no longer depends on the /register route being there:

https://github.com/stormpath/stormpath-framework-tck/pull/289

File: src/test/groovy/com/stormpath/tck/forgot/ChangePasswordIT.groovy (hunk 6 above)
Comment (dogeared) [inline src/test/groovy/com/stormpath/tck/forgot/ChangePasswordIT.groovy]: The above link is for `register`. @edjiang did point out that we decided to remove a password confirmation from the framework spec in the past. However, it's not currently specified one way or the other. It's not worth code changes in the Java integrations IMHO to remove the confirm password field for resetting your password. @omgitstom - would you please weigh in on this? It's holding up @edjiang's willingness to merge this PR.

**Comment by jarias** [comment]
@dogeared Did you look at this before implementing the CSRF support?

https://github.com/rest-assured/rest-assured/wiki/Usage#csrf

**Comment by edjiang** [comment]
Awesome! So glad we can get this merged in =]

**Pull request merged**